scripts/md-gen/compatibility-matrix/main.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/juju/version/v2"
+)
+
+type minVersionInfo struct {
+	minAgentVersion, minClientVersion string
+	minTerraformVersion               string
+	minPylibjujuVersion               string
+}
+
+func main() {
+	outputDir := mustEnv("DOCS_DIR") // directory to write output to
+
+	releaseTags := getReleasesFromJujuRepo()
+	minVersionsInfo := map[string]minVersionInfo{}
+	for _, tag := range releaseTags {
+		version := getVersionFromTag(tag)
+		minAgentVersion, minClientVersion := getMinAgentClientVersions(tag)
+		minVersionsInfo[version] = minVersionInfo{
+			minAgentVersion:  minAgentVersion,
+			minClientVersion: minClientVersion,
+		}
+	}
+	outputMarkdown(minVersionsInfo, outputDir)
+}
+
+func getTagsFromJujuRepo() []string {
+	resp, err := http.Get("https://api.github.com/repos/juju/juju/git/matching-refs/tags")
+	check(err)
+	defer resp.Body.Close()
+
+	var tagInfo []map[string]any
+	err = json.NewDecoder(resp.Body).Decode(&tagInfo)
+	check(err)
+
+	var tags []string
+	for _, info := range tagInfo {
+		tag, _ := strings.CutPrefix(info["ref"].(string), "refs/tags/")
+		tags = append(tags, tag)
+	}
+	return tags
+}
+
+func getReleasesFromJujuRepo() []string {
+	resp, err := http.Get("https://api.github.com/repos/juju/juju/releases")
+	check(err)
+	defer resp.Body.Close()
+
+	var releaseInfo []map[string]any
+	err = json.NewDecoder(resp.Body).Decode(&releaseInfo)
+	check(err)
+
+	var tags []string
+	for _, info := range releaseInfo {
+		tags = append(tags, info["tag_name"].(string))
+	}
+	return tags
+}
+
+func getVersionFromTag(tag string) (vers string) {
+	switch {
+	case strings.HasPrefix(tag, "juju-"):
+		vers, _ = strings.CutPrefix(tag, "juju-")
+	case strings.HasPrefix(tag, "v"):
+		vers, _ = strings.CutPrefix(tag, "v")
+	default:
+		panic(fmt.Sprintf("unknown tag format %q", tag))
+	}
+	version.MustParse(vers)
+	return
+}
+
+func getMinAgentClientVersions(tag string) (minAgentVersion, minClientVersion string) {
+	rawURL := fmt.Sprintf("https://raw.githubusercontent.com/juju/juju/%s/upgrades/upgradevalidation/version.go", tag)
```

Comment:
That doesn't live there for the 4.0 main branch. I move it to `/internal/`. I probably need to update it for 4.0 to ensure it only migrates from 3.6.

Reply:
right, I noticed that `MinClientVersions` doesn't exist in 4.0, and then I noticed that the value is not even used in 3.5 for example. So it probably doesn't make sense to put it in the table.

Is there anywhere in 3.5 that we gate which Juju client version can talk to which controller version?

Reply:
To define compatibility between juju versions (client, agent controller), then the only way to do that currently is to look at the facade versions from each and work out the intersection. That's harder than expected because in 2.9, we support everything from 0 to the valid facade version:

https://github.com/juju/juju/blob/737b09066686476cf72025dcef5466cdef3f5be3/api/facadeversions.go#L13-L128

I changed this in 3.x, so we dropped the concept of 0 facade (< juju 1.0) and implemented windowed facade versions that allow holes.

https://github.com/juju/juju/blob/fe493bf39c81dd93ca24b014f78f8249eb952701/api/facadeversions.go#L23-L142

You need to have a valid intersection between facadeversions, then you know it's supported.

Further, I would love to break these up into agent, client, and controller facade versions. Once we know that we can be more specific about what type of intersection you care about. Then the compatibility matrix becomes a lot more useful.